bin/tools.py

```diff
@@ -67,31 +67,37 @@
 # Make sure f-strings are supported.
 f'f-strings are not supported by your python version. You need at least python 3.6.'
 
-# Get the list of relevant problems.
-# Either use the problems.yaml, or check the existence of problem.yaml and sort
-# by shortname.
-def get_problems():
-    def is_problem_directory(path):
-        return (path / 'problem.yaml').is_file()
-
-    contest = None
-    problem = None
-    level = None
+
+# A path is a problem directory if it contains a `problem.yaml` file.
+def is_problem_directory(path):
+    return (path / 'problem.yaml').is_file()
+
+
+# Changes the working directory to the root of the contest.
+# Returns the "level" of the current command (either 'problem' or 'problemset')
+# and, if `level == 'problem'`, the directory of the problem.
+def change_directory():
+    problem_dir = None
+    level = 'problemset'
     if config.args.contest:
-        contest = config.args.contest.resolve()
-        os.chdir(contest)
-        level = 'problemset'
+        os.chdir(config.args.contest.resolve())
     if config.args.problem:
-        problem = config.args.problem.resolve()
+        problem_dir = config.args.problem.resolve()
         level = 'problem'
-        os.chdir(problem.parent)
+        os.chdir(problem_dir.parent)
     elif is_problem_directory(Path('.')):
-        problem = Path().cwd()
+        problem_dir = Path().cwd()
         level = 'problem'
         os.chdir('..')
-    else:
-        level = 'problemset'
 
+    config.level = level
+    return problem_dir
+
+
+# Get the list of relevant problems.
+# Either use the problems.yaml,
+# or check the existence of problem.yaml and sort by shortname.
+def get_problems(problem_dir):
     # We create one tmpdir per contest.
     h = hashlib.sha256(bytes(Path().cwd())).hexdigest()[-6:]
     tmpdir = Path(tempfile.gettempdir()) / ('bapctools_' + h)
```

Comment:
Maybe this should also change to `config.tmpdir`, like `config.level`. Not sure though.

Reply:
That's possible, but then the same would hold for `contest_name`, `problems`, and maybe even `problem` :stuck_out_tongue: Where would you draw the line between moving parameters to `config` and just passing them as parameters to the functions? In the other extreme, the global `config.level` (and maybe even `config.args`) should not exist, but instead be passed as a local parameter :stuck_out_tongue:

Reply:
Yeah fair enough. Also, I could see tmpdir change value within a single invocation, so let's just keep it as is.